### ISSUE TITLE
Bounds check fail/integer overflow in unicode_data::skip_search via cc when building with LTO

### ISSUE THREAD

**Issue by saethlin** [rust-lang/rust#82890]
```
RUSTFLAGS="-Cembed-bitcode=yes -Clto=fat -Cdebuginfo=2" cargo install cargo-make --force
```
```
  thread 'main' panicked at 'index out of bounds: the len is 31 but the index is 18446744073709551615', library/core/src/unicode/unicode_data.rs:82:62
```

I forgot the `--target=x86_64-unknown-linux-gnu` that would be required to make this actually work (build will eventually fail because LTO and proc-macro dylib crates). If I add the `--target`, I don't see any crashes. About half the time I run the `cargo install` invocation, I get a panic and the backtrace below (`-Cdebuginfo=2` for the backtrace, doesn't seem related to the crash). I do not see this panic if I swap out `-Clto=fat` for `-Copt-level=3`.

### Meta
`rustc --version --verbose`:
```
rustc 1.52.0-nightly (234781afe 2021-03-07)
binary: rustc
commit-hash: 234781afe33d3f339b002f85f948046d8476cfc9
commit-date: 2021-03-07
host: x86_64-unknown-linux-gnu
release: 1.52.0-nightly
LLVM version: 12.0.0
```

<details><summary>Backtrace</summary>
<p>

```
error: failed to run custom build command for `openssl-sys v0.9.60`

Caused by:
  process didn't exit successfully: `/tmp/cargo-installjmgYde/release/build/openssl-sys-ad87746b93d0b377/build-script-main` (exit code: 101)
  --- stdout
  cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
  OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
  OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_DIR
  X86_64_UNKNOWN_LINUX_GNU_OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_DIR
  OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=SYSROOT
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rustc-link-lib=ssl
  cargo:rustc-link-lib=crypto
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR

  --- stderr
  thread 'main' panicked at 'index out of bounds: the len is 31 but the index is 18446744073709551615', library/core/src/unicode/unicode_data.rs:82:62
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/std/src/panicking.rs:493:5
     1: core::panicking::panic_fmt
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/panicking.rs:92:14
     2: core::panicking::panic_bounds_check
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/panicking.rs:69:5
     3: core::unicode::unicode_data::skip_search::{{closure}}
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/unicode/unicode_data.rs:82:62
     4: core::option::Option<T>::map
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/option.rs:487:29
     5: core::unicode::unicode_data::skip_search
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/unicode/unicode_data.rs:82:9
     6: core::unicode::unicode_data::grapheme_extend::lookup
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/unicode/unicode_data.rs:305:9
     7: core::char::methods::<impl char>::is_grapheme_extended
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/char/methods.rs:859:9
     8: core::char::methods::<impl char>::escape_debug_ext
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/char/methods.rs:415:46
     9: core::char::methods::<impl char>::escape_debug
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/char/methods.rs:461:9
    10: <str as core::fmt::Debug>::fmt
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/fmt/mod.rs:2057:23
    11: <alloc::string::String as core::fmt::Debug>::fmt
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/alloc/src/string.rs:1966:9
    12: <&T as core::fmt::Debug>::fmt
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/fmt/mod.rs:2010:62
    13: core::fmt::builders::DebugTuple::field::{{closure}}
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/fmt/builders.rs:345:17
    14: core::result::Result<T,E>::and_then
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/result.rs:704:22
    15: core::fmt::builders::DebugTuple::field
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/fmt/builders.rs:332:23
    16: <core::option::Option<T> as core::fmt::Debug>::fmt
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/option.rs:158:48
    17: core::fmt::write
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/fmt/mod.rs:1092:17
    18: core::fmt::Write::write_fmt
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/fmt/mod.rs:182:9
    19: alloc::fmt::format
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/alloc/src/fmt.rs:578:5
    20: cc::Build::getenv
               at /home/ben/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.67/src/lib.rs:2553:21
    21: cc::Build::getenv_unwrap
               at /home/ben/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.67/src/lib.rs:2559:15
    22: cc::Build::get_opt_level
               at /home/ben/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.67/src/lib.rs:2520:24
    23: cc::Build::try_get_compiler
               at /home/ben/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.67/src/lib.rs:1291:25
    24: cc::Build::try_expand
               at /home/ben/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.67/src/lib.rs:1217:24
    25: build_script_main::validate_headers
               at ./build/main.rs:138:26
    26: build_script_main::find_normal::try_pkg_config
               at ./build/find_normal.rs:205:5
    27: build_script_main::find_normal::find_openssl_dir
               at ./build/find_normal.rs:75:5
    28: build_script_main::find_normal::get_openssl::{{closure}}
               at ./build/find_normal.rs:15:68
    29: core::option::Option<T>::unwrap_or_else
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/option.rs:427:21
    30: build_script_main::find_normal::get_openssl
               at ./build/find_normal.rs:15:31
    31: build_script_main::find_openssl
               at ./build/main.rs:55:5
    32: build_script_main::main
               at ./build/main.rs:63:34
    33: core::ops::function::FnOnce::call_once
               at /rustc/234781afe33d3f339b002f85f948046d8476cfc9/library/core/src/ops/function.rs:227:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
warning: build failed, waiting for other jobs to finish...
error: failed to compile `cargo-make v0.32.13`, intermediate artifacts can be found at `/tmp/cargo-installjmgYde`
```

</p>
</details>

**Comment by slightlyoutofphase** [comment]
I can confirm that this issue also exists when building in release mode for the `x86_64-pc-windows-msvc` target for at least the past couple of nightlies.

For me, it fails with exactly the same `unicode_data.rs` error and points to [this line of code in my crate](https://github.com/slightlyoutofphase/staticvec/blob/ad812d18da514fbb551aea09e37279c602742508/src/string/mod.rs#L26), which oddly is not even "code" really but just the opening triple-backticks of a previously-working doctest example.

**Comment by moxian** [comment]
I've spent some time trying to repro @slightlyoutofphase's error. Turns out rustdoc does weird things to tests... Anyhow
```toml
[profile.release]
opt-level = 0  # important!
lto = true 
codegen-units = 1

[dependencies]
staticvec = {git = "https://github.com/slightlyoutofphase/staticvec.git", rev="ad812d18da514fbb551aea09e37279c602742508"}
```
```rust
use staticvec::{StaticString, StringError};

#[derive(Debug)]
pub struct User {
  pub username: StaticString<20>,
  pub role: StaticString<5>,
}

fn main() -> Result<(), StringError> {
  let user = User {
    username: StaticString::try_from_str("user")?,
    role: StaticString::try_from_str("admin")?,
  };
  println!("{:?}", user);
  Ok(())
}
```
and
```rust
> cargo +nightly run --release
    Finished release [unoptimized] target(s) in 0.29s
     Running `target\release\rust-issue-82890.exe`
thread 'main' panicked at 'index out of bounds: the len is 31 but the index is 18446744073709551615', library\core\src\unicode\unicode_data.rs:82:62
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
User { username: StaticString { array: "error: process didn't exit successfully: `target\release\rust-issue-82890.exe` (exit code: 101)
```

**Comment by moxian** [comment]
Minified, kinda (as in: it is now just a single self-contained file):

<details>

```rust
#![feature(
    const_mut_refs,
    const_panic,
    const_ptr_is_null,
    const_raw_ptr_deref,
    const_slice_from_raw_parts,
)]

use core::char::DecodeUtf16Error;
use core::fmt::{self, Debug, Formatter};
use core::mem::{size_of, MaybeUninit};
use core::str::Utf8Error;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct CapacityError<const N: usize>;

/// This enum represents several different possible "error states" that may be encountered
/// while using a [`StaticString`](crate::string::StaticString).
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum StringError {
    /// Indicates a failed conversion from a `u8` slice to a
    /// [`StaticString`](crate::string::StaticString).
    Utf8(Utf8Error),
    /// Indicates a failed conversion from a `u16` slice to a
    /// [`StaticString`](crate::string::StaticString).
    Utf16(DecodeUtf16Error),
    /// Indicates an attempted access of an invalid UTF-8 character index.
    NotCharBoundary,
    /// Indicates an out-of-bounds indexed access of a [`StaticString`](crate::string::StaticString)
    /// instance.
    OutOfBounds,
}
impl<const N: usize> From<CapacityError<N>> for StringError {
    #[inline(always)]
    fn from(_err: CapacityError<N>) -> Self {
        Self::OutOfBounds
    }
}

/// A [`Vec`](alloc::vec::Vec)-like struct (mostly directly API-compatible where it can be)
/// implemented with const generics around an array of fixed `N` capacity.
struct StaticVec<T, const N: usize> {
    // We create this field in an uninitialized state, and write to it element-wise as needed
    // via pointer methods. At no time should `assume_init` *ever* be called through it.
    data: MaybeUninit<[T; N]>,
    // The constant `N` parameter (and thus the total span of `data`) represent capacity for us,
    // while the field below represents, as its name suggests, the current length of a StaticVec
    // (that is, the current number of "live" elements) just as is the case for a regular `Vec`.
    length: usize,
}

/// A local (identically written) `const fn` version of `slice::from_raw_parts`.
#[inline(always)]
pub(crate) const fn slice_from_raw_parts<'a, T>(data: *const T, length: usize) -> &'a [T] {
    debug_assert!(
        /*
        is_aligned_and_not_null(data),
        "Attempted to create an unaligned or null slice!"
        */
        // See comment starting at line 154 for more info about what's going on here.
        !data.is_null(),
        "Attempted to create a null slice!"
    );
    debug_assert!(
        size_of::<T>().saturating_mul(length) <= isize::MAX as usize,
        "Attempted to create a slice covering at least half of the address space!"
    );
    unsafe { &*core::ptr::slice_from_raw_parts(data, length) }
}

impl<T, const N: usize> StaticVec<T, N> {
    #[inline(always)]
    const fn new() -> Self {
        Self {
            data: Self::new_data_uninit(),
            length: 0,
        }
    }
    #[inline(always)]
    pub(crate) const fn new_data_uninit() -> MaybeUninit<[T; N]> {
        MaybeUninit::uninit()
    }
    #[inline(always)]
    const fn remaining_capacity(&self) -> usize {
        N - self.length
    }
    #[inline(always)]
    const fn len(&self) -> usize {
        self.length
    }
    #[inline(always)]
    pub(crate) const fn first_ptr_mut(this: &mut MaybeUninit<[T; N]>) -> *mut T {
        this as *mut MaybeUninit<[T; N]> as *mut T
    }
    #[inline(always)]
    const fn as_mut_ptr(&mut self) -> *mut T {
        Self::first_ptr_mut(&mut self.data)
    }
    #[inline(always)]
    unsafe fn mut_ptr_at_unchecked(&mut self, index: usize) -> *mut T {
        // We check against `N` as opposed to `length` in our debug assertion here, as these
        // `_unchecked` versions of `ptr_at` and `mut_ptr_at` are primarily intended for
        // initialization-related purposes (and used extensively that way internally throughout the
        // crate.)
        debug_assert!(
            index <= N,
            "In `StaticVec::mut_ptr_at_unchecked`, provided index {} must be within `0..={}`!",
            index,
            N
        );
        self.as_mut_ptr().add(index)
    }
    #[inline(always)]
    unsafe fn set_len(&mut self, new_len: usize) {
        // Most of the `unsafe` functions in this crate that are heavily used internally
        // have debug-build-only assertions where it's useful.
        debug_assert!(
            new_len <= N,
            "In `StaticVec::set_len`, provided length {} exceeds the maximum capacity of {}!",
            new_len,
            N
        );
        self.length = new_len;
    }
    #[inline(always)]
    pub(crate) const fn first_ptr(this: &MaybeUninit<[T; N]>) -> *const T {
        this as *const MaybeUninit<[T; N]> as *const T
    }
    #[inline(always)]
    const fn as_ptr(&self) -> *const T {
        Self::first_ptr(&self.data)
    }

    #[inline(always)]
    const fn as_slice(&self) -> &[T] {
        // Safety: `self.as_ptr()` is a pointer to an array for which the first `length`
        // elements are guaranteed to be initialized. Therefore this is a valid slice.
        slice_from_raw_parts(self.as_ptr(), self.length)
    }
}

pub(crate) struct StaticString<const N: usize> {
    pub(crate) vec: StaticVec<u8, N>,
}

impl<const N: usize> StaticString<N> {
    #[inline(always)]
    pub(crate) const fn new() -> Self {
        Self {
            vec: StaticVec::new(),
        }
    }
    #[inline(always)]
    pub(crate) fn try_from_str<S: AsRef<str>>(string: S) -> Result<Self, CapacityError<N>> {
        let mut res = Self::new();
        res.try_push_str(string)?;
        Ok(res)
    }

    #[inline(always)]
    pub(crate) const fn len(&self) -> usize {
        self.vec.len()
    }

    #[inline(always)]
    pub(crate) const fn remaining_capacity(&self) -> usize {
        self.vec.remaining_capacity()
    }

    #[inline(always)]
    pub(crate) unsafe fn push_str_unchecked(&mut self, string: &str) {
        let string_length = string.len();
        debug_assert!(string_length <= self.remaining_capacity());
        let old_length = self.len();
        let dest = self.vec.mut_ptr_at_unchecked(old_length);
        string.as_ptr().copy_to_nonoverlapping(dest, string_length);
        self.vec.set_len(old_length + string_length);
    }

    #[inline(always)]
    pub(crate) fn try_push_str<S: AsRef<str>>(
        &mut self,
        string: S,
    ) -> Result<(), CapacityError<N>> {
        let string_ref = string.as_ref();
        match self.vec.remaining_capacity() < string_ref.len() {
            false => {
                unsafe { self.push_str_unchecked(string_ref) };
                Ok(())
            }
            true => Err(CapacityError {}),
        }
    }
    #[inline(always)]
    pub(crate) const fn as_str(&self) -> &str {
        unsafe { &*(self.as_bytes() as *const [u8] as *const str) }
    }

    #[inline(always)]
    pub(crate) const fn as_bytes(&self) -> &[u8] {
        self.vec.as_slice()
    }
}

impl<const N: usize> Debug for StaticString<N> {
    #[inline(always)]
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        f.debug_struct("StaticString")
            .field("array", &self.as_str())
            .field("size", &self.len())
            .finish()
    }
}

#[derive(Debug)]
struct User {
    username: StaticString<20>,
    role: StaticString<5>,
}

fn main() -> Result<(), StringError> {
    let user = User {
        username: StaticString::try_from_str("user")?,
        role: StaticString::try_from_str("admin")?,
    };
    println!("{:?}", user);
    Ok(())
}
```
</details>

**Comment by moxian** [comment]
The code does use a bit of unsafe, but at a glance looks roughly sane (as in: I've spent 5 minutes staring at it and see no issues). So this _might_ be unsoundness somewhere in rustc/core?

@rustbot label: +T-prioritize

**Comment by moxian** [comment]
update:

```rust
fn main() {
    let data_ref: &[u8] = &[0u8];

    let username_str: &str = unsafe { &*(data_ref as *const [u8] as *const str) };
    println!("{:?}", username_str);
}
```

**Comment by nagisa** [comment]
Is this reproduced after https://github.com/rust-lang/rust/pull/83030?

**Comment by slightlyoutofphase** [comment]
@moxian Nice investigation! I will note that that the error was happening for me with a regular release profile for which `opt-level` is `3`, and not `0`, though.

**Comment by moxian** [comment]
@slightlyoutofphase when rustdoc runs doctests, it unsets `opt-level` but happens to keep `lto`. Which was very confusing trying to figure out - compiling the code with `opt-level=3` was stubbornly refusing to repro despite "clearly" having that in the original Cargo.toml.

**Comment by slightlyoutofphase** [comment]
@moxian Ah, I didn't know that about doctests. I don't think I have something that demonstrates anything different then - in fact it seemed extra strange because I did try copying that code out into a normal program with the same release profile, and in that context the error failed to surface.

So it seems like your narrowing it down to `opt-level = 0` is specifically correct for sure.

**Comment by danielhenrymantilla** [comment]
FWIW, here is a [Playground repro](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f4395182d1fea60773e91bd5cdc9accf)

**Comment by Aaron1011** [comment]
This can be reproduced with `format!` instead of `println!`

**Comment by danielhenrymantilla** [comment]
```text
stack backtrace:
   0: rust_begin_unwind
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/panicking.rs:92:14
   2: core::panicking::panic_bounds_check
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/panicking.rs:69:5
   3: core::unicode::unicode_data::skip_search::{{closure}}
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/unicode/unicode_data.rs:82:62
   4: core::option::Option<T>::map
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/option.rs:487:29
   5: core::unicode::unicode_data::skip_search
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/unicode/unicode_data.rs:82:9
   6: core::unicode::unicode_data::grapheme_extend::lookup
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/unicode/unicode_data.rs:305:9
   7: core::char::methods::<impl char>::is_grapheme_extended
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/char/methods.rs:859:9
   8: core::char::methods::<impl char>::escape_debug_ext
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/char/methods.rs:415:46
   9: core::char::methods::<impl char>::escape_debug
             at /rustc/b3e19a221e63dcffdef87e12eadf1f36a8b90295/library/core/src/char/methods.rs:461:9
  10: <str as core::fmt::Debug>::fmt
```

**Comment by moxian** [comment]
Minimized to not have any `fmt` machinery at all:

```rust
pub fn main() {
    let username = "Steve";    

    for (_, c) in username.char_indices() {
        let _esc = c.escape_debug();
    }
}
```

**Comment by nagisa** [comment]
Some investigation I did inside a debugger so far:

https://github.com/rust-lang/rust/blob/acca818928654807ed3bc1ce0e97df118f8716c8/library/core/src/unicode/unicode_data.rs#L69-L73

gets computed as 0 initially, which then has 1 subtracted from it:

https://github.com/rust-lang/rust/blob/acca818928654807ed3bc1ce0e97df118f8716c8/library/core/src/unicode/unicode_data.rs#L81-L82

resulting in a `-1` stored in the `A` register, which is then stored on a stack and a byte of `A` is set to `1` (that is, unsigned overflow has occurred)

```asm
   0x0000557f25342a61 <+449>:   sub    $0x1,%rax
   0x0000557f25342a65 <+453>:   mov    %rax,0x78(%rsp)
   0x0000557f25342a6a <+458>:   setb   %al
```

However the next immediate instruction is...

```asm
   0x0000557f25342a6d <+461>:   xor    %eax,%eax
   0x0000557f25342a6f <+463>:   mov    %eax,0x84(%rsp)
   0x0000557f25342a76 <+470>:   jb     0x557f25342ab6 <_ZN4core7unicode12unicode_data15grapheme_extend6lookup17hb68bf130fc9c0d7eE+534>
```

which is a short for… clear the 32 bits of `A` register, in an apparent attempt to clear a stack slot? This, unfortunately ends up entirely ignoring the fact an overflow has occurred (`xor` also clears the flags).

This suggests to me that there's an issue with register liveness tracking (the register holding the overflow flag ends up being killed early), and since in the underlying LLVM-IR these are represented as `{ T, i1 }`, there's going to be an `extractvalue` somewhere in there. As thus the bug sounds extremely similar to the issue fixed in https://github.com/rust-lang/rust/pull/83030 (https://bugs.llvm.org/show_bug.cgi?id=49467), except this time it apparently crops up through LTO.

**Comment by nikic** [comment]
IR for the relevant function: https://gist.github.com/nikic/74a6461a4c09cdf2bf6b2f6ce9d8582a Reproduces under `llc -O0`.

Initial machine instrs produced by FastISel already look broken:

```
  %14:gr64 = PHI %47:gr64, %bb.7, %56:gr64, %bb.8
  %64:gr64_nosp = SUB64ri8 %11:gr64_nosp(tied-def 0), 1, implicit-def $eflags
  %65:gr8 = SETCCr 2, implicit $eflags
  %60:gr32 = MOV32r0 implicit-def $eflags
  JCC_1 %bb.13, 2, implicit $eflags
```

**Comment by nikic** [comment]
Upstream report: https://bugs.llvm.org/show_bug.cgi?id=49587

**Comment by JohnTitor** [comment]
Assigning `P-medium` [as discussed as part of the Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2382890.20Bounds.20check.20fail.2Finteger.20overflow.20in.20unicode_data.3A.3A.E2.80.A6) and removing `I-prioritize`.

**Comment by nikic** [comment]
Upstream fix: https://github.com/llvm/llvm-project/commit/7669455df49e6fc8ae7d9f4bd4ee95bb20e7eb6e

**Comment by jhwgh1968** [comment]
I see that the LLVM submodule has been updated a couple times since the upstream fix, and there are a couple of PRs currently in queue for other LLVM issues.

Any idea when this fix will get into nightly? Or did it already?

**Comment by nikic** [comment]
Should be fixed by #84230.

**Comment by Mark-Simulacrum** [comment]
Fixed in 1.52 by https://github.com/rust-lang/rust/pull/84271 as well, backporting #84230.
